src/librustc_data_structures/smallvec.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! A vector type that allows storing small values directly on the stack.
+
+use std::iter::FromIterator;
+use std::ptr::drop_in_place;
+use std::ops::{Deref, DerefMut};
+use std::slice;
+use std::fmt;
+
+pub struct SmallVec8<T> {
+    count: usize,
+    values: [ManuallyDrop<T>; 8]
+}
+
+impl<T> SmallVec8<T> {
+    fn new() -> Self {
+        SmallVec8 {
+            count: 0,
+            values: [
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+                ManuallyDrop { empty: () },
+            ],
+        }
+    }
+}
+
+impl<T> fmt::Debug for SmallVec8<T>
+    where T: fmt::Debug {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self[..].fmt(f)
+    }
+}
+
+impl<T> Deref for SmallVec8<T> {
+    type Target = [T];
+    fn deref(&self) -> &Self::Target {
+        unsafe {
+            slice::from_raw_parts(self.values.as_ptr() as *const T, self.count)
+        }
+    }
+}
+
+impl<T> DerefMut for SmallVec8<T> {
+    fn deref_mut(&mut self) -> &mut [T] {
+        unsafe {
+            slice::from_raw_parts_mut(self.values.as_mut_ptr() as *mut T, self.count)
+        }
+    }
+}
+
+// FIXME: This should use repr(transparent) from rust-lang/rfcs#1758.
+#[allow(unions_with_drop_fields)]
+union ManuallyDrop<T> {
+    value: T,
+    #[allow(drop)]
+    empty: (),
+}
+
+impl<T> Drop for SmallVec8<T> {
+    fn drop(&mut self) {
+        unsafe {
+            drop_in_place(&mut self.values[..self.count]);
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum MaybeSmallVec8<T> {
+    Small(SmallVec8<T>),
+    Large(Vec<T>)
+}
+
+impl<T> Deref for MaybeSmallVec8<T> {
```

Comment:
Initially I didn't even expect this type to be visible outside of ty::substs. I probably need to write up a rationale for it but it's the best you can do without compromising.

It's definitely not general purpose, it only serves as a preallocation buffer for the common case of type or parameter lists.
I wouldn't use this type for anything else, not without const generics.

OTOH the SmallVector in libsyntax *also* has a limited purpose! It should truly be renamed OneOrMany or something like that, it is an optimization over T -> Vec<T> during AST folding - that is, some kinds of nodes can expand into several more, but most of the time it's just 1 and allocating would be wasteful.

**Neither** is suited for general-purpose SVO, save for SmallVector in cases where you really only have at most one element most of the time.